Services/PredefinedValuesProductAttributeService.cs

```diff
@@ -0,0 +1,47 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using OrchardCore.Commerce.Abstractions;
+using OrchardCore.Commerce.Models;
+using OrchardCore.Commerce.Settings;
+using OrchardCore.ContentManagement;
+
+namespace OrchardCore.Commerce.Services
+{
+    public class PredefinedValuesProductAttributeService : IPredefinedValuesProductAttributeService
+    {
+        private readonly IProductAttributeService _productAttributeService;
+
+        public PredefinedValuesProductAttributeService(IProductAttributeService productAttributeService)
+        {
+            _productAttributeService = productAttributeService;
+        }
+
+        public IEnumerable<IEnumerable<string>> GetProductAttributesPredefinedValues(ContentItem product)
+        {
+            return _productAttributeService.GetProductAttributeFields(product)
+                .Where(x => x.Settings is TextProductAttributeFieldSettings textSettings && textSettings.RestrictToPredefinedValues)
+                .Select(x => (x.Settings as TextProductAttributeFieldSettings).PredefinedValues.ToList())
+                .ToList();
+        }
+
+        public IEnumerable<string> GetProductAttributesCombinations(ContentItem product)
+        {
+            return CartesianProduct(GetProductAttributesPredefinedValues(product))
+                .Select(x => string.Join("-", x));
+        }
+
+        private IEnumerable<IEnumerable<T>> CartesianProduct<T>(IEnumerable<IEnumerable<T>> sequences)
+        {
+            IEnumerable<IEnumerable<T>> emptyProduct = new[] { Enumerable.Empty<T>() };
+            return sequences.Aggregate(
```

Comment:
I think you can optimize this in terms of allocations by using `SelectMany` and by doing a unique join of a flat enumeration.

Reply:
We cant use `SelectMany` here b/c we don't know how much sequences we have on compile time: [https://ericlippert.com/2010/06/28/computing-a-cartesian-product-with-linq/](url)